Plt 6711 Add client side file encryption

tejaycar opened the pull request:
#### Summary
Adds client side encryption for files.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6711

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [X ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ X] Has UI changes
- [ X] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

tejaycar commented:
@jwilander This is the PR for client side encryption since Minio doesn't support SSE for copy yet.  You mind taking a look for me?

jwilander commented on app/diagnostics.go:
```diff
@@ -241,17 +241,19 @@ func trackConfig() {
+		"driver_name":         		utils.Cfg.FileSettings.DriverName,
+		"amazon_s3_ssl":       		*utils.Cfg.FileSettings.AmazonS3SSL,
+		"amazon_s3_signv2":    		*utils.Cfg.FileSettings.AmazonS3SignV2,
+		"at_rest_encryption_key":	*utils.Cfg.FileSettings.AtRestEncryptKey,
```
Is there a reason you added the key here? This data is sent back to us for tracking if these config settings are being used, I don't think you want to send us your encryption key

tejaycar replied:
Honestly, I wasn't sure what this was doing and figured it was more likely you'd catch it if it wasn't needed than if I left it out and it was needed.  That said, would you like a boolean value here so you can track if encryption is being used?

jwilander replied:
A boolean would be useful yes

jwilander commented on app/file.go:
```diff
@@ -71,14 +72,27 @@ func ReadFile(path string) ([]byte, *model.AppError) {
 		secure := *utils.Cfg.FileSettings.AmazonS3SSL
-		signV2 := *utils.Cfg.FileSettings.AmazonS3SignV2
+		encrypted := *utils.Cfg.FileSettings.AtRestEncryptKey != ""
+    signV2 := *utils.Cfg.FileSettings.AmazonS3SignV2
```
Can you run `gofmt` on your code? It looks like the indentation is off

tejaycar replied:
oddly, when I run `make check-style` it doesn't report any issues, but when I run gofmt directly it does.  I'll update it.

cpanato replied:
if you use sublime, you can install the gofmt plugin that helps a lot :)

jwilander commented on model/config.go:
```diff
@@ -1383,6 +1389,10 @@ func (o *Config) IsValid() *AppError {
 		return NewLocAppError("Config.IsValid", "model.config.is_valid.file_salt.app_error", nil, "")
 	}
 
+	if len(*o.FileSettings.AtRestEncryptKey) > 0 && len(*o.FileSettings.AtRestEncryptKey) < 32 {
```
Is there a reason we're limiting they key length to 32?

tejaycar replied:
we're not.  It must be either 0 or >= 32.  0 = do not encrypt.  I required at least 32 because that's what was required for the at rest encryption key for the database.

jwilander commented on cprofile.out:
```diff
@@ -0,0 +1,15585 @@
+mode: count
```
I think this file was added unintentionally

tejaycar replied:
yup

tejaycar commented:
@jwilander  I incorporated your inputs, and also split this into two fields
EnableClientSideEncryption
and
ClientSideEncryptionKey

That seems more consistent with what you guys have elsewhere and less confusing if we add in server side encryption in the future (per PLT-6648)

tejaycar commented:
@jwilander sorry for the long delay.  We now have a successful build incorporating your requested changes as well as improving the naming of the variable and adding a boolean variable for turning it on and off.

jwilander commented on app/file.go:
```diff
@@ -71,14 +72,27 @@ func ReadFile(path string) ([]byte, *model.AppError) {
+		} else {
+			minioObject, err = s3Clnt.GetObject(bucket, path)
+		}
+		//	defer minioObject.Close()
```
Can you remove this comment if it's not needed?

jwilander commented on model/config.go:
```diff
@@ -1545,6 +1566,9 @@ func (o *Config) Sanitize() {
 	}
+	if len(*o.FileSettings.ClientSideEncryptionKey) > 0 {
+		*o.FileSettings.ClientSideEncryptionKey = FAKE_SETTING
+	}
```
You will also need to desanitize it in `utils/config.go`

tejaycar replied:
@jwilander  done

coreyhulen commented on app/file.go:
```diff
@@ -71,14 +72,26 @@ func ReadFile(path string) ([]byte, *model.AppError) {
 		}
 		bucket := utils.Cfg.FileSettings.AmazonS3Bucket
-		minioObject, err := s3Clnt.GetObject(bucket, path)
-		defer minioObject.Close()
```
The defer close is still needed otherwise you're leaking resources.

tejaycar replied:
@coreyhulen  my lack of Go experience is a problem here.  The minio GetEncryptedObject returns an io.Reader object, which doesn't have a close method.  The GetObject method that was used before returns an "Object" instead.

And, as I look more closely here, I'm not really sure what type should be declared on line 83 as I don't know what common type I can use for both of those methods.  Any ideas?

tejaycar replied:
@coreyhulen  talked to the minio guys, and it looks like they have a bug.  so we're dependent on https://github.com/minio/minio-go/pull/699 which I'm told should be merged in today.  I've got the updated code ready to go as soon as they merge in their fix.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

coreyhulen commented:
@jasonblais we should test this.  We can use a CI server S3 bucket if needed (or bladekick).  Just make sure to change the root password.

esethna commented:
Hi @tejaycar, highly appreciate your work on this.

As per our [release process](https://docs.mattermost.com/process/release-process.html) we lock down third party dependencies in the early part of the release cycle. Due to the changes in this PR, we will not be able to take it in v3.10, as we are only a few days away from cutting the first release candidate.

We would be happy to take this change in the version after next, releasing July 16th.

jasonblais commented:
@esethna I think this might be needed for a customer, can you confirm with @coreyhulen or @lfbrock?

jasonblais commented:
@tejaycar can you help resolve the build failure and I'll set up a test server for it?

tejaycar commented:
@jasonblais  I'm the customer this is for :).  I've got the code ready to push, but it is waiting on minio/minio-go#699 before the code will work.  They said it should be merged in today, but so far no luck.  Should I push the code as is anyway?  If so, how do I kick off a build test once the dependent library is updated?

tejaycar commented:
@jasonblais it required updating the minio code in the vendor directory, but I've got a clean build now that I believe is ready to go.  Not entirely sure the interplay between "we lock down dependencies early" and "needed for a customer" (especially a brand new customer).  I guess I'll just keep my fingers crossed.

mattermod commented:
Spinmint test server created at: http://i-044008b8ed31d3418.spinmint.com:8065/pr6525

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-044008b8ed31d3418

hmhealey commented on model/config.go:
```diff
@@ -489,6 +491,21 @@ func (o *Config) SetDefaults() {
+			*o.FileSettings.ClientSideEncryptionKey = NewRandomString(32)
+		}
+	} else {
+		o.FileSettings.ClientSideEncryptionKey = new(string)
```
Do we want to clear the key if encryption is disabled? I think we usually just leave it as is as long as its not configured. Maybe change the `else` part to `else if o.FileSettings.ClientSideEncryptionKey == nil`

tejaycar replied:
sounds good

tejaycar replied:
@hmhealey I added that change, but it looks like master is currently broken.  Will this pr automatically re-build when master is fixed, or do I need to somehow kick it back off?

hmhealey replied:
I'll trigger Jenkins to rebuild it. If that doesn't fix it, you might just need to rebase/merge onto master

hmhealey commented on webapp/components/admin_console/storage_settings.jsx:
```diff
@@ -258,6 +262,43 @@ export default class StorageSettings extends AdminSettings {
+                    label={
+                        <FormattedMessage
+                            id='admin.file.enableClientSideEncryptionTitle'
+                            defaultMessage='Encrypt Files at the Mattermost server'
```
@jasonblais We need some review on this text. We should at least make it clear that this only affects Amazon S3 file storage, and probably group it with the other S3 settings. Should that be done after this PR is merged, or do we want to do it now?

tejaycar replied:
@hmhealey  it is set to only show if the S3 driver is selected.  It should also work with S3 compatible systems like Minio.

hmhealey replied:
I think it's just disabled but still visible when local file storage is selected. I think it's fine to still display it when local file storage is enabled, but it should be clearly grouped with the other S3 settings

tejaycar commented:
@coreyhulen  how do I get someone to update minio?  I've got the code for this PR ready to go, but it requires a version of minio-go that includes minio/minio-go#699

jasonblais commented:
@tejaycar Can you help share this information with `christopher` on pre-release (e.g. in the [Developers channel](https://pre-release.mattermost.com/core/channels/developers)) and he can include the Minio update on the next 3rd-library dependancy update PR. The PR will be submitted soon now that 3.10 is out.

tejaycar commented:
I recommend we close this in favor of #6467

jasonblais commented:
@coreyhulen @tejaycar Is this something we want to add support for, or is SSE sufficient?

coreyhulen commented:
I believe that SSE is sufficient for now, but 0/5 on if we want this too.

tejaycar commented:
I wouldn't use it, personally.

On Aug 24, 2017 1:03 PM, "Corey Hulen" <notifications@github.com> wrote:

> I believe that SSE is sufficient for now, but 0/5 on if we want this too.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattermost/platform/pull/6525#issuecomment-324727371>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADqwzKwCtH6gR5_gqiRjFEG0U0iVAW1cks5sbcj-gaJpZM4Nq6pe>
> .
>

jasonblais commented:
Okay, I'll close this PR for now then. We can reopen later if needed.
